src/gen10_hevc_enc_common.c

```diff
@@ -0,0 +1,1398 @@
+        frame_info->gop_num_b[2] = 0;
+
+        if (gen10_hevc_enc_get_relocation_flag(seq_param,
+                                               &frame_info->last_seq_param))
+            frame_info->reallocate_flag = 1;
+        else
+            frame_info->reallocate_flag = 0;
+
+        memcpy(&frame_info->last_seq_param, seq_param, sizeof(*seq_param));
+    } else
+        frame_info->reallocate_flag = 0;
+}
+
+static int
+hevc_find_skipemulcnt(uint8_t *buf, int bits_length)
+{
+    int skip_cnt = 0, i = 0;
+
+    if ((bits_length >> 3) < 6)
+        return 0;
+
+    for (i = 0; i < 3; i++)
+        if (buf[i] != 0)
+            break;
+
+    if (i > 1) {
+        if (buf[i] == 1)
+            skip_cnt = i + 3;
+    }
+
+    return skip_cnt;
+}
+
+static void
+gen10_hevc_enc_insert_object(VADriverContextP ctx,
+                             struct intel_batchbuffer *batch,
+                             uint8_t *header_data,
+                             int length_in_bits,
+                             int end_of_slice_flag,
+                             int last_header_flag,
+                             int emulation_flag,
+                             int skip_emulation_bytes)
+{
+    gen10_hcp_pak_insert_object_param insert_param;
+
+    memset(&insert_param, 0, sizeof(insert_param));
+
+    insert_param.dw1.bits.end_of_slice_flag = end_of_slice_flag;
+    insert_param.dw1.bits.last_header_flag = last_header_flag;
+    insert_param.dw1.bits.emulation_flag = emulation_flag;
+
+    if (emulation_flag) {
+        if (skip_emulation_bytes)
+            insert_param.dw1.bits.skip_emulation_bytes = skip_emulation_bytes;
+        else
+            insert_param.dw1.bits.skip_emulation_bytes = hevc_find_skipemulcnt((uint8_t *)header_data,
+                                                                               length_in_bits);
+    }
+
+    insert_param.dw1.bits.data_bits_in_last_dw = length_in_bits & 0x1f;
+    if (insert_param.dw1.bits.data_bits_in_last_dw == 0)
+        insert_param.dw1.bits.data_bits_in_last_dw = 32;
+
+    insert_param.inline_payload_ptr = header_data;
+    insert_param.inline_payload_bits = length_in_bits;
+
+    gen10_hcp_pak_insert_object(ctx, batch, &insert_param);
+}
+
+void
+gen10_hevc_enc_insert_packed_header(VADriverContextP ctx,
+                                    struct encode_state *encode_state,
+                                    struct intel_encoder_context *encoder_context,
+                                    struct intel_batchbuffer *batch)
+{
+    VAEncPackedHeaderParameterBuffer *param = NULL;
+    uint8_t *header_data = NULL;
+    uint32_t length_in_bits = 0;
+    int packed_type = 0;
+    int idx = 0, idx_offset = 0;
+    int i = 0;
+
+    for (i = 0; i < 4; i++) {
+        idx_offset = 0;
+        switch (i) {
+        case 0:
+            packed_type = VAEncPackedHeaderHEVC_VPS;
+            break;
+        case 1:
+            packed_type = VAEncPackedHeaderHEVC_VPS;
```

Comment:
VPS is repeated; perhaps SPS?

Reply:
yes, here VPS means SPS. but I'd like to keep it for SPS should be inserted just after VPS and they has the same header type VAEncPackedHeaderSequence. only idx_offset is 1, this header means SPS

Reply:
@fhvwy , thanks for all your comments.  Now, we don't have plan to release source about those shader codes.